src/typecheck/lsp.rs

```diff
@@ -0,0 +1,76 @@
+//! Special typing analysis utils that should be done when we run the type checker in LSP mode
```

Comment:
General concern with this PR and feature: why is this extra inference only done for LSP?
E.g. the exact same problem happens with the REPL:
```
> cat lib.ncl
{ name = "lib" }
> nickel repl
nickel> :typecheck import "lib.ncl"
Ok: Dyn
```

I am also concerned with the LSP typechecker diverging from Nickel's typechecker for a few reasons:
* double maintenance
* risk of introducing bugs (there is code duplication, so e.g. if someone modifies the `Term::Num(_)` case in `typecheck/mod.rs`, one MUST also do the same change in `typecheck/lsp.rs`
* unclear separation of concerns: the LSP is a different code base, but now we have LSP-specific code in the Nickel codebase

I am holding off on reviewing the PR until this concern is answered

Reply:
Well, I originally implemented this on the type-checker, but there were issues regarding Nickel's semantics. See [this](https://github.com/tweag/nickel/pull/878#issuecomment-1282592098). We just need this to be able to perform completion for imports and stdlib terms. 

For those concerns, having good tests for the LSP might help, but you're right.

Reply:
All this context - why it's done in the type checking code even though it's LSP-specific; why the code is duplicated rather than enhanced; why we don't want to typecheck imports in regular Nickel; etc. - is currently only present in the Github comment. It is not present in the code or documentation of this PR, meaning that future readers will have no idea why this is done this way.

If we want to stick with this design, then the rationales for this choice have to be documented. We must leave trails for future readers to understand the past decisions.

I don't have a strong opinion on whether it's directly documented in the code, or if the code instead refers to an [Architectural Decision Record](https://github.com/joelparkerhenderson/architecture-decision-record) where this is documented.